javascript/ql/src/semmle/javascript/frameworks/EventEmitter.qll

```diff
@@ -0,0 +1,225 @@
+      result = pred.track(t2, t)
+      or
+      // invocation of a chainable method
+      exists(DataFlow::MethodCallNode mcn |
+        mcn = pred.getAMethodCall(EventEmitter::chainableMethod()) and
+        // exclude getter versions
+        exists(mcn.getAnArgument()) and
+        result = mcn and
+        t = t2.continue()
+      )
+    )
+  }
+
+  /**
+   * An object that implements the EventEmitter API.
+   * Extending this class does nothing, its mostly to indicate intent.
+   * The magic only happens when extending EventRegistration::Range and EventDispatch::Range.
+   */
+  abstract class Range extends DataFlow::Node { }
+
+  /**
+   * An NodeJS EventEmitter instance.
+   * Events dispatched on this EventEmitter will be handled by event handlers registered on this EventEmitter.
+   * (That is opposed to e.g. SocketIO, which implements the same interface, but where events cross object boundaries).
+   */
+  abstract class NodeJSEventEmitter extends Range {
+    /**
+     * Get a Node that refers to a NodeJS EventEmitter instance.
+     */
+    DataFlow::SourceNode ref() { result = trackEventEmitter(this) }
+  }
+
+  private class ImportedNodeJSEventEmitter extends NodeJSEventEmitter {
+    ImportedNodeJSEventEmitter() {
+      exists(DataFlow::SourceNode clazz |
+        clazz = DataFlow::moduleImport("events") or
+        clazz = DataFlow::moduleMember("events", "EventEmitter")
+      |
+        this = clazz.getAnInstantiation()
+      )
+    }
+  }
+}
+
+/**
+ * An EventEmitter instance that implements the NodeJS EventEmitter API.
+ * Extend EventEmitter::Range to mark something as being an EventEmitter.
+ */
+class EventEmitter extends DataFlow::Node {
+  EventEmitter::Range range;
+
+  EventEmitter() { this = range }
+}
+
+/**
+ * A registration of an event handler on an EventEmitter.
+ */
+class EventRegistration extends DataFlow::Node {
+  EventRegistration::Range range;
+
+  EventRegistration() { this = range }
+
+  /** Gets the EventEmitter that the event handler is registered on. */
+  final EventEmitter getEmitter() { result = range.getEmitter() }
+
+  /** Gets the name of the channel if possible. */
+  string getChannel() { result = range.getChannel() }
+
+  /** Gets the `i`th parameter in the event handler. */
+  DataFlow::Node getReceivedItem(int i) { result = range.getReceivedItem(i) }
+
+  /**
+   * Gets a value that is returned by the event handler.
+   * The default implementation is that no value can be returned.
+   */
+  DataFlow::Node getAReturnedValue() { result = range.getAReturnedValue() }
+
+  /**
+   * Get a dispatch that this event handler can return a value to. 
+   * The default implementation is that there exists no such dispatch.
+   */
+  EventDispatch getAReturnDispatch() { result = range.getAReturnDispatch() }
+}
+
+module EventRegistration {
+  /**
+   * A registration of an event handler on an EventEmitter.
+   * The default implementation assumes that `this` is a DataFlow::InvokeNode where the
```

Comment:
Hmm, do we ever provide default implementations in the `Range` part of the Range pattern?

Will we run into the usual inheritance problems we had before the introduction of the range pattern by providing default implementations here? I do not think so, but it is worth a sanity check from a second pair of eyes (@max-schaefer?).

Reply:
See https://github.com/Semmle/ql/blob/f1f69ef85dff1d0bb324e13fdd81b2149afb4826/javascript/ql/src/semmle/javascript/dataflow/Nodes.qll#L654-L683

Reply:
Yeah, it seems cleaner to pull those default implementations out into an abstract subclass and have subclasses of `::Range` that want to use those default implementations extend that abstract subclass instead of `::Range` itself.